articles/0002-api-sets.mdx

```diff
@@ -0,0 +1,189 @@
+
+At the same time, they wanted the flexibility to implement functions wherever needed.
+
+# Analysis
+
+Let's start by taking a look at the good old `kernel32.dll`. We can see lots of function with `Stub` suffix, what's going on here? 
+
+![alt](/article/0002/kernel32-stubs.png)
+
+Let's pick one at random and find out. I'll choose `GetLastError` without any particular reason behind it. 
+
+We can see that it's exported as `GetLastError`, without the `Stub` suffix and that it does nothing besides calling `__imp_GetLastError`. The name didn't lie, it's really just a stub.
+
+![alt](/article/0002/kernel32-get-last-error-stub.png)
+
+What's `__imp_GetLastError` though? We haven't seen that function yet. Let's take a peek.
+
+![alt](/article/0002/kernel32-api-ms-win-core-errorhandling-imports.png)
+
+It seems to be part of imports from `api-ms-win-core-errorhandling-l1-1-0.dll`. There's no actual implementation of `GetLastError` in `kernel32.dll`. Seems like it's one of those functions they wanted to move somewhere else.
+
+Well, what are we waiting for? Let's take a look at `api-ms-win-core-errorhandling-l1-1-0` but first we need to find it. 
+
+Where is it? This particular one is located in `C:\Windows\System32\downlevel`, which is where most of those `api-ms-` libraries are found, though it's not a strict rule. Some may be located elsewhere.
+
+![alt](/article/0002/c-windows-system32-downlevel.png)
+
+Also, I'll mention as a side note that DLL loading order still applies here. When an application requests a DLL, Windows follows a specific search order to locate it. This includes checking the application's directory first, then system directories, and finally any paths listed in the environment variables. So in the case below, Windows will first try to search for the relevant DLLs in the application directory.
+
+![alt](/article/0002/league-api-sets.png)
+
+The first thing we'll see after opening `api-ms-win-core-errorhandling-l1-1-0` is that there are no symbols whatsoever. That's concerning.
+
+![alt](/article/0002/api-ms-win-core-errorhandling-l1-1-0-symbols.png)
+
+But there are exports! Let's take a look at them.
+
+![alt](/article/0002/api-ms-win-core-errorhandling-l1-1-0-exports.png)
+
+Uh oh. There's something called a `forwarder` and they all seem to be pointing back at `kernel32`! That's infinite recursion, isn't it? What the heck is going on here?
+
+It turns out those libraries are special. They are not supposed to be treated like most other libraries. You'll never resolve functions by doing so. Instead, there's a mapping of those empty API set libraries to actual libraries with real implementations.
+
+This mapping is created at boot according to `Windows/System32/apisetschema.dll` and additional schemas defined in `Computer/HKEY_LOCAL_MACHINE/SYSTEM/CurrentControlSet/Control/Session Manager/ApiSetSchemaExtensions`.
+
+![alt](/article/0002/schema-extensions-in-registry.png)
```

Comment:
fajnie byłoby wspomnieć co to za soft jak się pojawia jakiś po raz pierwszy - ktoś w temacie może kojarzyć, ale ktoś spoza bańki to trochę rng